src/rdkafka.h

```diff
@@ -4205,6 +4372,947 @@ rd_kafka_interceptor_add_on_request_sent (
+
+/**
+ * @brief Create a new AdminOptions object.
+ *
+ *        The options object is not modified by the Admin API request APIs,
+ *        (e.g. CreateTopics) and may be reused for multiple calls.
+ *
+ * @param rk Client instance.
+ * @param for_api Specifies what Admin API this AdminOptions object will be used
+ *                for, which will enforce what AdminOptions_set_..() calls may
+ *                be used based on the API, causing unsupported set..() calls
+ *                to fail.
+ *                Specifying NULL disables the enforcement.
+ *                Valid values are: CreateTopics, DeleteTopics,
+ *                                  CreatePartitions,
+ *                                  AlterConfigs, DescribeConfigs.
+ *
+ * @returns a new AdminOptions object (which must be freed with
+ *          rd_kafka_AdminOptions_destroy()), or NULL if \p for_api was set to
+ *          an unknown API name.
+ */
+RD_EXPORT rd_kafka_AdminOptions_t *
+rd_kafka_AdminOptions_new (rd_kafka_t *rk, const char *for_api);
+
+/**
+ * @brief Destroy a AdminOptions object.
+ */
+RD_EXPORT void rd_kafka_AdminOptions_destroy (rd_kafka_AdminOptions_t *options);
+
+
+/**
+ * @brief Sets the overall request timeout, including broker lookup,
+ *        request transmission, operation time on broker, and response.
+ *
+ * @param timeout_ms Timeout in milliseconds, use -1 for indefinite timeout.
+ *                   Defaults to `socket.timeout.ms`.
+ *
+ * @returns RD_KAFKA_RESP_ERR_NO_ERROR on success, or
+ *          RD_KAFKA_RESP_ERR__INVALID_ARG if timeout was out of range in which
+ *          case an error string will be written \p errstr.
+ *
+ * @remark This option is valid for all Admin API requests.
+ */
+RD_EXPORT rd_kafka_resp_err_t
+rd_kafka_AdminOptions_set_request_timeout (rd_kafka_AdminOptions_t *options,
+                                           int timeout_ms,
+                                           char *errstr, size_t errstr_size);
+
+
+/**
+ * @brief Sets the broker's operation timeout, such as the timeout for
+ *        CreateTopics to complete the creation of topics on the controller
+ *        before returning a result to the application.
+ *
+ * CreateTopics: values <= 0 will return immediately after triggering topic
+ * creation, while > 0 will wait this long for topic creation to propagate
+ * in cluster. Default: 0.
+ *
+ * DeleteTopics: same semantics as CreateTopics.
+ * CreatePartitions: same semantics as CreateTopics.
+ *
+ *
+ * @param timeout_ms Timeout in milliseconds.
+ *
+ * @returns RD_KAFKA_RESP_ERR_NO_ERROR on success, or
+ *          RD_KAFKA_RESP_ERR__INVALID_ARG if timeout was out of range in which
+ *          case an error string will be written \p errstr.
+ *
+ * @remark This option is valid for CreateTopics, DeleteTopics and 
+ *         CreatePartitions.
+ */
+RD_EXPORT rd_kafka_resp_err_t
+rd_kafka_AdminOptions_set_operation_timeout (rd_kafka_AdminOptions_t *options,
```

Comment:
I did not want to maintain an additional type for this specific use only, but I guess you are right.

Reply:
Primarily that, but also because some of the Admin operations have broker-side timeouts ( e.g., createTopics and deleteTopics) which tells the broker how long it may block waiting for an operation to propagate throughout the cluster.
But from a user's perspective we also want a request-level timeout that includes broker lookup, connection, request time, and the operation time.
In a previous PR I tried to encode both these timeouts in a single value but it was ugly and did not go down well with the Mr Math.

We could call the operation timeout just `timeout` as in the Java API, but with two timeouts in play I believe this is too ambigious, thus the operation_timeout and request_timeout.